### PR TITLE
68 test and improve precision of straight driving and turning

### PR DESCRIPTION
**Pull request by MikaelFangel**
### Changes
- Change the driving function to drive with the gyro
- Added PID to both rotate and drive
- Changed the proto file to provide information about the constants for PID
- Removed the stop motor check because it causes the motors to stop unevenly
- Removed old drive and rotate methods
- Refactored the go modules

### Other 
This only solves the precision of the driving and rotating function. For the precision of distance, checkout the branch related to #38.

This also solves the problem discussed with making a command queue as all functions block until the task has finished.

### CODE REVIEW

File: remote/src/main/java/RobotController.java [hunk 1]
```diff
@@ -30,40 +33,66 @@ public void stopController() throws InterruptedException {
...
     /**
-     * Makes the robot drive straight either forward or backwards for the distance given as argument
+     * Makes the robot drive straight either forward or backwards by using the gyro
      * @param distance Positive values in cm for forward and negative for backwards
      * @throws RuntimeException if the robot was not reached
      */
-    public void driveStraight(double distance) throws RuntimeException {
-       MultipleMotors motorsRequest = createMultipleMotorRequest(Type.l, new MotorPair(OutPort.A, DEFAULT_SPEED),
-               new MotorPair(OutPort.D, DEFAULT_SPEED));
+    public void driveWGyro(double distance) throws RuntimeException {
```
Comment (Greve2001) [inline remote/src/main/java/RobotController.java:40]: The name driveWGyro does not make sense to an outsider. I really dont know if the W stand for 'with'.

File: remote/src/main/java/RobotController.java [hunk 2]
```diff
@@ -30,40 +33,66 @@ public void stopController() throws InterruptedException {
...
-        DriveRequest driveRequest = DriveRequest.newBuilder()
+        DrivePIDRequest driveRequest = DrivePIDRequest.newBuilder()
                 .setMotors(motorsRequest)
-                .setDistance((float) distance)
-                .setSpeed(DEFAULT_SPEED)
+                .setDistance((float) distance) // Note: Currently not used on the robot
+                .setSpeed(speed) // This speed worked well, other speeds could be researched
+                .setKp(0.5f)
+                .setKi(0.25f)
+                .setKd(0.1f)
```
Comment (Greve2001) [inline remote/src/main/java/RobotController.java:51]: What are these values??? :confused:

File: remote/src/main/java/RobotController.java (hunk 1 above)
Comment (MikaelFangel) [reply remote/src/main/java/RobotController.java:40]: w/ is a common abbreviation in the English language for with.

File: remote/src/main/java/RobotController.java [hunk 3]
```diff
@@ -30,40 +33,66 @@ public void stopController() throws InterruptedException {
...
-            CLIENT.rotate(rotateRequest);
+            CLIENT.rotateWGyro(rotateRequest);
+        }
+        catch (RuntimeException e) {
+            System.out.println(e.getMessage());
+        }
+    }
+
+    /**
+     * Recalibrates the
```
Comment (Greve2001) [inline remote/src/main/java/RobotController.java:91]: Can you end this JavaDoc

File: robot/main.go [hunk 4]
```diff
@@ -57,63 +49,94 @@ func main() {
...
+	// Change speed value if distance is negative
+	speed := -int(in.Speed)
+	if in.Speed < 0 {
+		speed *= -1
+	}
 
-//----------------------------------------------------------------------------------------------------------------------
-// Global functions
-//----------------------------------------------------------------------------------------------------------------------
+	// TODO: For testing purposes run until i < 20, but should be until length to target <= 0. Research RPC client streaming
```
Comment (Greve2001) [inline robot/main.go:84]: Is this todo not implemented?

File: remote/src/main/java/RobotController.java [hunk 5]
```diff
@@ -30,40 +33,66 @@ public void stopController() throws InterruptedException {
...
+                .setKp(0.5f)
+                .setKi(0.25f)
+                .setKd(0.1f)
                 .build();
 
-        CLIENT.drive(driveRequest);
+        CLIENT.driveWGyro(driveRequest);
+    }
+
+    public void stopMotors() {
```
[the remark this answers is not in the document]
Comment (MikaelFangel) [reply remote/src/main/java/RobotController.java:60]: Added.

File: robot/main.go [hunk 6]
```diff
@@ -57,63 +49,94 @@ func main() {
...
+		// "D term", trying to predict next error
+		turn := (kp * gyroErr) + (ki * integral) + (kd * derivative)
+
+		for _, request := range in.GetMotors().GetMotor() {
+			motor, err := util.GetMotorHandle(request.GetMotorPort().String(), request.GetMotorType().String())
+			if err != nil {
+				return &pBuff.StatusReply{ReplyMessage: false}, err
+			}
+
+			// TODO: Should ramp be set for all iterations or only first and last? I found that the these values worked well
```
Comment (Greve2001) [inline robot/main.go:109]: Should this comment be removed, or at least be rewritten as one of the possibilities are implemented?

File: robot/main.go [hunk 7]
```diff
@@ -57,63 +49,94 @@ func main() {
...
+			motor.SetRampUpSetpoint(3 * time.Second)
+			motor.SetRampDownSetpoint(3 * time.Second)
```
Comment (Greve2001) [inline robot/main.go:111]: Is 3 a constant factor that we use for this? Should the time sent not just be 3 times larger?

File: robot/main.go [hunk 8]
```diff
@@ -57,63 +49,94 @@ func main() {
...
 
-//----------------------------------------------------------------------------------------------------------------------
-// Global functions
-//----------------------------------------------------------------------------------------------------------------------
+	// TODO: For testing purposes run until i < 20, but should be until length to target <= 0. Research RPC client streaming
+	gyro, err := util.GetSensor(pBuff.InPort_in1.String(), pBuff.Sensor_gyro.String())
+	if err != nil {
+		return nil, err
+	}
+	for i := 0; i < int(in.Distance); i++ {
```
Comment (Greve2001) [inline robot/main.go:89]: Could you please describe in comments what this big code block should do. It comes a bit out of nowhere

File: robot/main.go [hunk 9]
```diff
@@ -57,63 +49,94 @@ func main() {
...
-			return &pBuff.StatusReply{ReplyMessage: false}, err
+		derivative := gyroErr - lastError
+		lastError = gyroErr
+
+		// "P term", how much we want to change the motors' power in proportion with the error
+		// "I term", the running sum of errors to correct for
+		// "D term", trying to predict next error
+		turn := (kp * gyroErr) + (ki * integral) + (kd * derivative)
+
+		for _, request := range in.GetMotors().GetMotor() {
```
Comment (Greve2001) [inline robot/main.go:103]: What does this for loop do?

File: robot/main.go [hunk 10]
```diff
@@ -133,90 +156,94 @@ func (s *motorServer) StopMotors(_ context.Context, in *pBuff.MultipleMotors) (*
...
+	lastError := 0.0
+	target := 0.0
 
-// Drive Makes the robot drive straight forward or backward with a speed and distance specified client side
-func (s *motorServer) Drive(_ context.Context, in *pBuff.DriveRequest) (*pBuff.StatusReply, error) {
 	var motorRequests []motorRequest
+	gyro, _ := util.GetSensor(pBuff.InPort_in1.String(), "GYRO")
+	gyroVal, _ := gyro.Value(0)
+	gyroValF, _ := strconv.ParseFloat(gyroVal, 64)
+	for math.Abs(gyroValF) != math.Abs(float64(in.Degrees)) {
```
Comment (Greve2001) [inline robot/main.go]: What does this for loop do?

File: robot/main.go [hunk 11]
```diff
@@ -133,90 +156,94 @@ func (s *motorServer) StopMotors(_ context.Context, in *pBuff.MultipleMotors) (*
...
 
-	return &pBuff.StatusReply{ReplyMessage: true}, nil
+	switch {
+	case power > 50:
+		power = 50
+	case power < -50:
+		power = -50
+	}
+
+	return power
```
Comment (Greve2001) [inline robot/main.go:242]: Please describe why we set a maxfor the power

**Review by Greve2001 — CHANGES_REQUESTED** [review]
The biggest issue i have is that many logical decisions are unclear as to what they achieve and are not communicated to the reader.

File: robot/main.go (hunk 11 above)
Comment (MikaelFangel) [reply robot/main.go:242]: It is already described on line 228.

**Pull request merged**